apiserver/facades/agent/uniter/storage.go

```diff
@@ -406,3 +415,54 @@ func accessUnitTag(tag string, canAccess func(names.Tag) bool) (names.UnitTag, e
 	}
 	return u, nil
 }
+
+// watchStorageAttachment returns a state.NotifyWatcher that reacts to changes
+// to the VolumeAttachmentInfo or FilesystemAttachmentInfo corresponding to the
+// tags specified.
+func watchStorageAttachment(
+	st storageInterface,
+	stVolume storageVolumeInterface,
+	stFile storageFilesystemInterface,
+	storageTag names.StorageTag,
+	machineTag names.MachineTag,
+	unitTag names.UnitTag,
+) (state.NotifyWatcher, error) {
+	storageInstance, err := st.StorageInstance(storageTag)
+	if err != nil {
+		return nil, errors.Annotate(err, "getting storage instance")
+	}
+	var watchers []state.NotifyWatcher
+	switch storageInstance.Kind() {
+	case state.StorageKindBlock:
+		volume, err := stVolume.StorageInstanceVolume(storageTag)
+		if err != nil {
+			return nil, errors.Annotate(err, "getting storage volume")
+		}
+		// We need to watch both the volume attachment, and the
+		// machine's block devices. A volume attachment's block
+		// device could change (most likely, become present).
+		watchers = []state.NotifyWatcher{
+			stVolume.WatchVolumeAttachment(machineTag, volume.VolumeTag()),
+			// TODO(axw) 2015-09-30 #1501203
+			// We should filter the events to only those relevant
+			// to the volume attachment. This means we would need
+			// to either start th block device watcher after we
+			// have provisioned the volume attachment (cleaner?),
+			// or have the filter ignore changes until the volume
+			// attachment is provisioned.
+			stVolume.WatchBlockDevices(machineTag),
+		}
+	case state.StorageKindFilesystem:
+		filesystem, err := stFile.StorageInstanceFilesystem(storageTag)
```

Comment:
Couldn't stFile be nil here? (I think maybe this couldn't happen, but it's probably not safe to assume it here.)

Reply:
done